### PR TITLE
XSI-312 CA-289145: Close socket if error occurs when connecting

### PR DESCRIPTION
**Pull request by lindig**
Backport of 65d985d

Signed-off-by: Yang Qian <yang.qian@citrix.com>
Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

### CODE REVIEW

File: lib/open_uri.ml [hunk 1]
```diff
@@ -15,36 +15,43 @@
  *
  *)
 
+open Xapi_stdext_pervasives.Pervasiveext
+
 let handle_socket f s =
```
Comment (edwintorok) [inline lib/open_uri.ml]: Is handle_socket a noop now?

File: lib/posix_channel.ml [hunk 2]
```diff
@@ -196,8 +196,14 @@ let receive protocols =
   | V4V_proxy(_, _) -> assert false (* weight is 0 above *)
   | TCP_proxy(ip, port) ->
     let s = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-    Unix.connect s (Unix.ADDR_INET(Unix.inet_addr_of_string ip, port));
-    s
+    begin
+      try
+        Unix.connect s (Unix.ADDR_INET(Unix.inet_addr_of_string ip, port));
+        s
+      with e ->
```
Comment (edwintorok) [inline lib/posix_channel.ml]: This likely looses the backtrace here

Comment (lindig) [reply lib/posix_channel.ml]: This is a straight backport by cherry-picking. Do we want to do more?

Comment (edwintorok) [reply lib/posix_channel.ml]: Lets backport this as is to Havana, and improve it on master.

**Pull request merged**